src/libraries/System.Security.Cryptography.Cose/tests/CoseKeyTests.SignVerify.cs

```diff
@@ -0,0 +1,585 @@
+            Assert.True(signerImplementation.Verify(KeyManager, keyId, payload, signature));
+
+            // we try different key
+            CoseTestKey differentKey = KeyManager.GetDifferentKey(keyId);
+            Assert.False(signerImplementation.Verify(KeyManager, differentKey.Id, payload, signature));
+
+            // we try bad signature (or bad key with good signature, same thing)
+            Assert.False(signerImplementation.Verify(BadKeyManager, keyId, payload, signature));
+
+            // we try fake payload
+            if (!signerImplementation.IsEmbedded)
+            {
+                // embedded ignore payload arg
+                byte[] fakePayload = Encoding.UTF8.GetBytes("Hello World 2");
+                Assert.False(signerImplementation.Verify(KeyManager, keyId, fakePayload, signature));
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(AllKeysAndMultiSignImplementations))]
+        public void TestSignVerifyMultiSignerAllAlgorithms(string[] keyIds, CoseTestMultiSign signerImplementation)
+        {
+            byte[] payload = Encoding.UTF8.GetBytes("Hello World");
+            byte[] signature = signerImplementation.Sign(KeyManager, keyIds, payload);
+            Assert.NotNull(signature);
+            Assert.True(signature.Length > 0);
+            Assert.True(signerImplementation.Verify(KeyManager, keyIds, payload, signature));
+
+            // we try fake signature
+            Assert.False(signerImplementation.Verify(BadKeyManager, keyIds, payload, signature));
+
+            // we try fake payload
+            if (!signerImplementation.IsEmbedded)
+            {
+                byte[] fakePayload = Encoding.UTF8.GetBytes("Hello World 2");
+                Assert.False(signerImplementation.Verify(KeyManager, keyIds, fakePayload, signature));
+            }
+        }
+
+        public static IEnumerable<object[]> AllKeysAndSign1Implementations()
+        {
+            foreach (string keyId in CoseTestKeyManager.GetAllKeyIds())
+            {
+                foreach (CoseTestSign1 sign1 in CoseTestSign1.GetImplementations())
+                {
+                    yield return [keyId, sign1];
+                }
+            }
+        }
+
+        public static IEnumerable<object[]> AllKeysAndMultiSignImplementations()
+        {
+            string[] keyIds = CoseTestKeyManager.GetAllKeyIds();
+            int[] nrOfKeys = [1, 2, 3, 3, 5];
+
+            for (int i = 0; i < keyIds.Length; i++)
+            {
+                string[] keysToTest = PickNKeys(nrOfKeys[i % nrOfKeys.Length], i, keyIds);
+                foreach (CoseTestMultiSign multiSign in CoseTestMultiSign.GetImplementations())
+                {
+                    yield return [keysToTest, multiSign];
+                }
+            }
+
+            static string[] PickNKeys(int n, int atIndex, string[] keys)
+            {
+                string[] ret = new string[n];
```

Comment:
Is it a problem if `n > keys.Length` (if the same key gets used twice?)?  If so, add an assert that it doesn't happen.